src/controllers/keyboard/keyboardeventfilter.cpp

```diff
@@ -156,15 +180,139 @@ QKeySequence KeyboardEventFilter::getKeySeq(QKeyEvent* e) {
 #endif
 }
 
-void KeyboardEventFilter::setKeyboardConfig(ConfigObject<ConfigValueKbd>* pKbdConfigObject) {
+void KeyboardEventFilter::setEnabled(bool enabled) {
+    m_enabled = enabled;
+    m_pConfig->set(ConfigKey("[Keyboard]", "Enabled"), ConfigValue(enabled));
+    emit shortcutsEnabled(enabled);
+}
+
+void KeyboardEventFilter::registerShortcutWidget(WBaseWidget* pWidget) {
+    m_widgets.append(pWidget);
+
+    // Tell widgets to reconstruct tooltips when option is toggled.
+    // WBaseWidget is not a QObject, need to use a lambda
+    connect(this,
+            &KeyboardEventFilter::shortcutsEnabled,
+            this,
+            [pWidget](bool enabled) {
+                pWidget->toggleKeyboardShortcutHints(enabled);
+            });
+}
+
+void KeyboardEventFilter::updateWidgetShortcuts() {
+    for (auto pWidget : m_widgets) {
+        QString shortcutHints;
+        const QList<std::pair<ConfigKey, QString>> keys = pWidget->getShortcutKeys();
+        QString keyString;
+        for (const auto& key : keys) {
+            keyString = m_pKbdConfig->getValueString(key.first);
+            if (!keyString.isEmpty()) {
+                shortcutHints.append(buildShortcutString(keyString, key.second));
+            }
+        }
+        // might be empty
+        pWidget->setShortcutTooltip(shortcutHints, m_enabled);
+    }
+}
+
+void KeyboardEventFilter::clearWidgets() {
+    disconnect(this, nullptr, nullptr, nullptr);
+    m_widgets.clear();
+}
+
+const QString KeyboardEventFilter::buildShortcutString(
+        const QString& shortcut, const QString& cmd) const {
+    if (shortcut.isEmpty()) {
+        return QString();
+    }
+
+    // translate shortcut to native text
+    QString nativeShortcut = QKeySequence(shortcut, QKeySequence::PortableText)
+                                     .toString(QKeySequence::NativeText);
+
+    QString shortcutTooltip;
+    shortcutTooltip += QObject::tr("Shortcut");
```

Comment:
`KeyboardEventFilter` already inherit `QObject`
```suggestion
    shortcutTooltip += tr("Shortcut");
```

Reply:
Yup. Leftover from when this was a namespace function.

Reply:
done.